src/Canvas.ts

```diff
@@ -0,0 +1,1015 @@
+  m12: number,
+  m21: number,
+  m22: number,
+  m31: number,
+  m32: number
+) => (ctx: CanvasRenderingContext2D) => IO.IO<CanvasRenderingContext2D> = (m11, m12, m21, m22, m31, m32) => (
+  ctx
+) => () => {
+  ctx.transform(m11, m12, m21, m22, m31, m32)
+  return ctx
+}
+
+/**
+ * Gets the current text alignment.
+ *
+ * @since 1.0.0
+ */
+export const getTextAlign: (ctx: CanvasRenderingContext2D) => IO.IO<TextAlign> = (ctx) => IO.of(ctx.textAlign)
+
+/**
+ * Sets the current text alignment.
+ *
+ * @since 1.0.0
+ */
+export const setTextAlign: (
+  textAlign: TextAlign
+) => (ctx: CanvasRenderingContext2D) => IO.IO<CanvasRenderingContext2D> = (ta) => (ctx) => () => {
+  ctx.textAlign = ta
+  return ctx
+}
+
+/**
+ * Gets the current font.
+ *
+ * @since 1.0.0
+ */
+export const getFont: (ctx: CanvasRenderingContext2D) => IO.IO<string> = (ctx) => IO.of(ctx.font)
+
+/**
+ * Sets the current font.
+ *
+ * @since 1.0.0
+ */
+export const setFont: (font: string) => (ctx: CanvasRenderingContext2D) => IO.IO<CanvasRenderingContext2D> = (f) => (
+  ctx
+) => () => {
+  ctx.font = f
+  return ctx
+}
+
+/**
+ * Render filled text.
+ *
+ * @since 1.0.0
+ */
+export const fillText: (
+  text: string,
+  x: number,
+  y: number,
+  maxWidth?: number
+) => (ctx: CanvasRenderingContext2D) => IO.IO<CanvasRenderingContext2D> = (t, x, y, mw) => (ctx) => () => {
+  if (typeof mw !== 'undefined') {
+    ctx.fillText(t, x, y, mw)
+  } else {
+    ctx.fillText(t, x, y)
+  }
+  return ctx
+}
+
+/**
+ * Render stroked text.
+ *
+ * @since 1.0.0
+ */
+export const strokeText: (
+  text: string,
+  x: number,
+  y: number,
+  maxWidth?: number
+) => (ctx: CanvasRenderingContext2D) => IO.IO<CanvasRenderingContext2D> = (t, x, y, mw) => (ctx) => () => {
+  if (typeof mw !== 'undefined') {
+    ctx.strokeText(t, x, y, mw)
+  } else {
+    ctx.strokeText(t, x, y)
+  }
+  return ctx
+}
+
+/**
+ * Get the text measurements for the specified text.
+ *
+ * @since 1.0.0
+ */
+export const measureText: (text: string) => (ctx: CanvasRenderingContext2D) => IO.IO<TextMetrics> = (t) => (ctx) =>
+  IO.of(ctx.measureText(t))
+
+/**
+ * Save the current canvas context.
+ *
+ * @since 1.0.0
+ */
+export const save: (ctx: CanvasRenderingContext2D) => IO.IO<CanvasRenderingContext2D> = (ctx) => () => {
+  ctx.save()
+  return ctx
+}
+
+/**
+ * Restore the previous canvas context.
+ *
+ * @since 1.0.0
+ */
+export const restore: (ctx: CanvasRenderingContext2D) => IO.IO<CanvasRenderingContext2D> = (ctx) => () => {
+  ctx.restore()
+  return ctx
+}
+
+/**
+ * A convenience function which allows for running an action while preserving the existing
+ * canvas context.
+ *
+ * @since 1.0.0
+ */
+export const withContext: <A>(
+  f: (ctx: CanvasRenderingContext2D) => IO.IO<A>
+) => (ctx: CanvasRenderingContext2D) => IO.IO<A> = (f) => (ctx) =>
+  pipe(
+    save(ctx),
+    IO.chain(() => f(ctx)),
+    IO.chain((a) =>
+      pipe(
+        restore(ctx),
+        IO.map(() => a)
+      )
+    )
```

Comment:
this pattern which is a bit cumbersome...

```ts
IO.chain((a) =>
  pipe(
    restore(ctx),
    IO.map(() => a)
  )
)
```

...can be replaced by `chainFirst`

```ts
IO.chainFirst(() => restore(ctx))
```

Reply:
I discover cleaner/better ways to write code with `fp-ts` every day! Thanks for the tip @gcanti. See https://github.com/gcanti/graphics-ts/pull/7/commits/375baca3c6c6687cd48c75c850d3efcbfc8ac92b